packages/core/src/event-store/postgres/store.ts

```diff
@@ -373,46 +376,132 @@ export class PostgresEventStore implements EventStore {
+              .values(mergedIntervalRows)
+              .execute();
+          }
+
+          return mergedIntervals;
+        });
+      })
+    );
 
-      return mergedIntervals;
+    const fragmentsWithIdx = fragments.map((f, idx) => ({ idx, ...f }));
+
+    const intervals = await this.db
+      .with(
+        "factoryFilterFragments(fragmentIndex, fragmentAddress, fragmentEventSelector, fragmentChildAddressLocation, fragmentTopic0, fragmentTopic1, fragmentTopic2, fragmentTopic3)",
+        () =>
+          sql`( values ${sql.join(
+            fragmentsWithIdx.map(
+              (f) =>
+                sql`( ${sql.val(f.idx)}, ${sql.val(f.address)}, ${sql.val(
+                  f.eventSelector
+                )}, ${sql.val(f.childAddressLocation)}, ${sql.val(
+                  f.topic0
+                )}, ${sql.val(f.topic1)}, ${sql.val(f.topic2)}, ${sql.val(
+                  f.topic3
+                )} )`
+            )
+          )} )`
+      )
+      .selectFrom("factoryLogFilterIntervals")
+      .leftJoin("factories", "factoryId", "factories.id")
+      .innerJoin("factoryFilterFragments", (join) => {
+        let baseJoin = join.on(({ or, cmpr }) =>
+          or([
+            cmpr("address", "is", null),
+            cmpr("fragmentAddress", "=", sql.ref("address")),
+          ])
+        );
+        baseJoin = join.on(({ or, cmpr }) =>
+          or([
+            cmpr("eventSelector", "is", null),
+            cmpr("fragmentEventSelector", "=", sql.ref("eventSelector")),
+          ])
+        );
+        baseJoin = join.on(({ or, cmpr }) =>
+          or([
+            cmpr("childAddressLocation", "is", null),
+            cmpr(
+              "fragmentChildAddressLocation",
+              "=",
+              sql.ref("childAddressLocation")
+            ),
+          ])
+        );
+        for (const idx_ of range(0, 4)) {
+          baseJoin = baseJoin.on(({ or, cmpr }) => {
+            const idx = idx_ as 0 | 1 | 2 | 3;
+            return or([
+              cmpr(`topic${idx}`, "is", null),
+              cmpr(`fragmentTopic${idx}`, "=", sql.ref(`topic${idx}`)),
+            ]);
+          });
+        }
+
+        return baseJoin;
+      })
+      .select(["fragmentIndex", "startBlock", "endBlock"])
+      .where("chainId", "=", chainId)
+      .execute();
+
+    const intervalsByFragment = intervals.reduce((acc, cur) => {
+      const { fragmentIndex, ...rest } = cur;
+      acc[fragmentIndex] ||= [];
+      acc[fragmentIndex].push({
+        startBlock: rest.startBlock,
+        endBlock: rest.endBlock,
+      });
+      return acc;
+    }, {} as Record<number, { startBlock: bigint; endBlock: bigint }[]>);
+
+    const fragmentIntervals = fragmentsWithIdx.map((f) => {
+      return (intervalsByFragment[f.idx] ?? []).map(
+        (r) =>
+          [Number(r.startBlock), Number(r.endBlock)] satisfies [number, number]
+      );
     });
+
+    return intervalIntersectionMany(fragmentIntervals);
```

Comment:
This code path was written before fragment IDs were a thing - I think we might be able to simplify this a bit by replacing this notion of fragment indexes with fragment IDs. Then, in a single pass, construct the `fragmentIntervals` array. Would be able to copy the same change in 4 places across the two stores.